contracts/flashloan/Instapool v2/main.sol

```diff
@@ -0,0 +1,193 @@
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import { Helper } from "./helpers.sol";
+
+import { 
+    IndexInterface,
+    ListInterface,
+    TokenInterface,
+    DSAInterface,
+    IAaveLending
+} from "./interfaces.sol";
+
+
+contract AaveFlashloaner is Helper {
+    using SafeERC20 for IERC20;
+
+    constructor(
+        address instaIndex_,
+        address wchainToken_,
+        address aaveLendingAddr_
+    ) Helper(instaIndex_, wchainToken_, aaveLendingAddr_) {}
+
+
+    event LogFlashLoan(
+        address indexed dsa,
+        address[] tokens,
+        uint256[] amounts
+    );
+
+    event LogWhitelistSig(bytes4 indexed sig, bool whitelist);
+
+    /**
+    * @dev Check if sig is whitelisted
+    * @param _sig bytes4
+    */
+    function checkWhitelisted(bytes4 _sig) public view returns(bool) {
+        return whitelistedSigs[_sig];
+    }
+
+    function calFee(uint[] memory amts_) external view returns (uint[] memory finalAmts_, uint[] memory premiums_, uint fee_) {
+        fee_ = aaveLending.FLASHLOAN_PREMIUM_TOTAL();
+        for (uint i = 0; i < amts_.length; i++) {
+            premiums_[i] = (amts_[i] * fee_) / 10000;
+            finalAmts_[i] = finalAmts_[i] + premiums_[i];
+        }
+    }
+
+    /**
+    * @dev Whitelists / Blacklists a given sig
+    * @param _sigs list of sigs
+    * @param _whitelist list of bools indicate whitelist/blacklist
+    */
+    function whitelistSigs(bytes4[] memory _sigs, bool[] memory _whitelist) public onlyOwner {
+        require(_sigs.length == _whitelist.length, "arr-lengths-unequal");
+        for (uint i = 0; i < _sigs.length; i++) {
+            require(!whitelistedSigs[_sigs[i]], "already-enabled");
+            whitelistedSigs[_sigs[i]] = _whitelist[i];
+            emit LogWhitelistSig(_sigs[i], _whitelist[i]);
+        }
+    }
+    
+    function executeOperation(
+        address[] calldata assets,
+        uint256[] calldata amounts,
+        uint256[] calldata premiums,
+        address initiator,
+        bytes calldata _data
+    ) external returns (bool) {
+        require(initiator == address(this), "not-same-sender");
+        require(msg.sender == aaveLendingAddr, "not-aave-sender");
+
+        uint _length = assets.length;
+        IERC20[] memory _tokenContracts = new IERC20[](_length);
+        bool[] memory isWChainToken = new bool[](_length);
+
+        CastData memory cd;
+        (cd.dsa, cd.callData) = abi.decode(
+            _data,
+            (address, bytes)
+        );
+
+        for (uint i = 0; i < _length; i++) {
+            _tokenContracts[i] = IERC20(assets[i]);
+            _tokenContracts[i].approve(aaveLendingAddr, amounts[i] + premiums[i]);
+            isWChainToken[i] = assets[i] == chainToken;
```

Comment:
It will never enter this logic. Since, we are changing chainToken to wchainToken

Reply:
So at the moment should we disallow WETH Flash borrow? Users can only flash borrow ETH & we can add the logic here w.r.t WETH as we changed the token address below.

Reply:
Yess. We won't WETH flash borrow and allow only ETH or vice verse.